`docs`: update supported features and reorganize documentation

rivos-eblot opened the pull request:
1. Move device specific information to dedicated doc files
2. Name device doc files after their implementation counterpart
3. Update the supported features for each OpenTitan machine
4. Factorize documentation so that doc is no longer duplicated (and often not in sync) across OpenTitan machines

pamaury commented:
I am not sure I entirely understand the change: that would create both a `ot_usbdev.md` and `usbdev.md` files for example?

rivos-eblot commented:
> I am not sure I entirely understand the change: that would create both a `ot_usbdev.md` and `usbdev.md` files for example?

No, `usbdev.md` becomes `ot_usbdev.md`, matching `ot_usbdev.c` implementation.

The original form of the documentation has been built from an accretion of pieces of information I wrote to use the OpenTitan machine and to track all the useful properties associated with each "driver". It became a mess as the emulation grew and some names became confusing between some tool-related documentation and the actual device documentation. 

Moreover, there are far too many duplicates between the different machines. It made sense earlier as EarlGrey and Darjeeling machine were diverging at lot, which is no longer the case where most of the code is shared (or could be, for the devices that remain split such as `ot_gpio`). It makes more sense to document each device with its own file, and index them in each machine documentation.

rivos-eblot committed:
[ot] docs/opentitan: darjeeling, earlgrey: add a bus recap section

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

AlexJones0 commented on docs/opentitan/ibex_hart.md:
```diff
@@ -0,0 +1,19 @@
+## Supported features
+
+* ePMP
+* Zbr ISA extension (crc32 instructions)
```
It might be nice to capture the unsupported features here as they could easily catch people out (from the top of my head: no NMIs, and no support for the non-standard RISC-V Bit-Manipulation draft version 0.93 Zbe/Zbf/Zbp/Zbt).

rivos-eblot replied:
Sure but I'm not sure at all of what is missing (only what is actually implemented) :)

rivos-eblot replied:
I've updated `ibex_hart.md`, not sure it represents a comprehensive list.

AlexJones0 commented on docs/opentitan/ot_darjeeling.md:
```diff
@@ -0,0 +1,147 @@
+## Supported version
+
+Please check out `hw/opentitan/ot_ref.log`
+
```
Nit: spacing

AlexJones0 commented on docs/opentitan/ot_darjeeling.md:
```diff
@@ -0,0 +1,147 @@
+  * HW bus config is ignored (SPI mode, speed, ...)
+* Timer
+* [UART](ot_uart.md)
+  * missing RX timeout, break support
```
RX Break support now exists via the CharDev, TX break does not (repeated for EG).

AlexJones0 commented on docs/opentitan/ot_darjeeling.md:
```diff
@@ -0,0 +1,147 @@
+### Near feature-complete devices
+
+* [AES](ot_aes.md)
+  * missing side-loading
```
As far as I'm aware AES, KMAC and OTBN sideloading are implemented? (repeated for EG)

AlexJones0 commented on docs/opentitan/ot_darjeeling.md:
```diff
@@ -0,0 +1,147 @@
+* [ROM controller](ot_rom_ctrl.md)
+* [SPI device](ot_spi_device.md)
+  * Flash mode supported
+  * TPM mode supported
```
"TPM mode supported, but shares a CS with flash/passthrough mode and so cannot be used together" or equivalent? (repeated in EG)

AlexJones0 commented on docs/opentitan/ot_darjeeling.md:
```diff
@@ -0,0 +1,147 @@
+features are implemented.
+
+* Analog Sensor Top
+
```
Nit: spacing

AlexJones0 commented on docs/opentitan/ot_earlgrey.md:
```diff
@@ -0,0 +1,137 @@
+  * noise source only (from host source)
+  * configurable clock sources
+* [GPIO](ot_gpio.md)
+  * Connections with pinmux not implemented (need to be ported from [Darjeeling](ot_darjeeling.md)
+    version)
```
Disclaimer: I know very little apart from the basics about (OpenTitan) Pinmux.

I have a question about this - I had a quick glance previously, but it looked like the Darjeeling Pinmux connections are to the DIO dedicated pad selects, whereas as far as I can tell Earlgrey has no DIO GPIO pad selects. Am I missing something else here - is there something that can be ported across to EG? Having pinmuxing for even just GPIOs would be quite helpful for us in testing Earlgrey.

rivos-eblot replied:
I thought both machines have DIO and MIO, but with very different number and assignment for each of them.
Maybe @jwnrt knows more about this, as he was working on this topic a couple of week ago I believe.

AlexJones0 replied:
Earlgrey does have both DIO and MIO, but I can only see MIO for the GPIO pins specifically on Earlgrey (in OpenTitan `top_earlgrey.h`) whereas Darjeeling seems to be using DIO for GPIO. But maybe I am just looking in the wrong place.

If you don't know it is no problem, I just thought it was worth asking :)

rivos-eblot replied:
Ok I understood it the wrong way. I was referring to the Pinmux device, not the GPIO subset, sorry.

rivos-eblot replied:
So back to the initial question: I think that the way the pinmux has been designed (and is used in our Darjeeling variant) should be ported to EarlGrey, in order to handle features such as:
- opendrain
- high-z
- weak pull-up, weak pull-down
and a combination of them (see `ibex_gpio.h`)

I think that the GPIO block could be factorized between EG and DJ.

AlexJones0 commented on docs/opentitan/ot_earlgrey.md:
```diff
@@ -0,0 +1,137 @@
+  * Interactions with other devices (such as the Reset Manager) are limited
+* [Reset Manager](ot_rstmgr.md)
+  * HW and SW reset requests are supported
+* [USB Device](ot_usbdev.md)
```
Fine to leave this as this for now but @pamaury this should probably be moved to "Partially implemented devices" and updated now that USBDev contains support for data transfers?

rivos-eblot replied:
Sure, done.

AlexJones0 commented on docs/opentitan/ot_usbdev.md:
```diff
@@ -0,0 +1,6 @@
+* `-chardev pty,id=usbdev-cmd` can be used to connect to the usbdev driver (commands).
+* `-chardev pty,id=usbdev-host` can be used to connect to the usbdev driver (host).
+
+Note: There is no USB device on the [Darjeeling machine](ot_darjeeling.md)
```
WRT @pamaury's comment - it looks like there is actually still a `usbdev.md` and `ot_usbdev.md`? Did you mean to move the existing USBDev documentation here?

rivos-eblot replied:
I thought it had been updated when I rebased, but it did not. This should be fixed in the next push.

AlexJones0 commented on docs/opentitan/ot_otp.md:
```diff
@@ -0,0 +1,8 @@
+  [`otptool.py`](otptool.md) tool.
+
+* on LC escalate reception, it is possible to early abort VM execution. Specify
+  `-global ot-otp-dj.fatal_escalate=true` to enable this feature.
```
Nit: this property needs to be updated to be top-agnostic (`ot-otp-eg` or `ot-otp-dj`).

AlexJones0 commented on docs/opentitan/ot_gpio.md:
```diff
@@ -7,6 +7,15 @@ There are two variations of the GPIO device: one for EarlGrey, one for Darjeelin
+When using multiple GPIO IPs, traces may become highly verbose, coming from multiple source.
+It is possible to limit the trace to a single GPIO IP, using the following option:
+
+`-global ot-gpio-dj.log_id=<ot_id>` where _ot_id_ is the OpenTitan identifier of the GPIO device.
```
Nit: add a note that this is not available on Earlgrey currently?

rivos-eblot committed:
[ot] docs/opentitan: docs: rename device documentation files with `ot_` prefix

1. It matches the actual device implementation file
2. It makes easier to spot device-related files from tools and other docs

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] docs/opentitan: darjeeling, earlgrey: rename machine documentation files

... to match implementation files

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] docs/opentitan: docs: move device documentation files, update supported features

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot commented:
Note sure if I can pass the lint-commit CI test: should I leave the commit messages as is, or add some `word:` to pass the check?

AlexJones0 reviewed (approved):
LGTM, thanks.

> Note sure if I can pass the lint-commit CI test: should I leave the commit messages as is, or add some word: to pass the check?

I think it could be possible to make the check conditionally allow this on `docs` commits but it might just be easier to add a redundant `docs/opentitan: docs: ... commit message ...` or similar (I think "docs" is reasonable since it is a general, holistic documentation update).

rivos-eblot commented:
> I think it could be possible to make the check conditionally allow this on `docs` commits but it might just be easier to add a redundant `docs/opentitan: docs: ... commit message ...` or similar

Ok, will do.
